internal/fs/dir_handle_test.go

```diff
@@ -0,0 +1,169 @@
+	bucket gcsx.SyncerBucket
+	clock  timeutil.SimulatedClock
+	dh     *fs.DirHandle
+}
+
+var _ SetUpInterface = &DirHandleTest{}
+var _ TearDownInterface = &DirHandleTest{}
+
+func init() {
+	RegisterTestSuite(&DirHandleTest{})
+}
+
+func (t *DirHandleTest) SetUp(ti *TestInfo) {
+	t.ctx = ti.Ctx
+	t.clock.SetTime(time.Date(2023, 6, 26, 15, 55, 0, 0, time.Local))
+	bucket := gcsfake.NewFakeBucket(&t.clock, fakeBucketName)
+	t.bucket = gcsx.NewSyncerBucket(
+		int64(appendThreshold), // Append threshold
+		tmpObjectPrefix,
+		bucket)
+}
+
+func TestAsync(t *testing.T) {
+	RunTests(t)
+}
+
+func (t *DirHandleTest) TearDown() {
+}
+
+// Create the inode and directory handle. No implicit dirs by default.
+func (t *DirHandleTest) createDirHandle(implicitDirs bool, enableNonexistentTypecache bool, dirInodeName string) {
+	in := inode.NewDirInode(
+		dirInodeID,
+		inode.NewDirName(inode.NewRootName(""), dirInodeName),
+		fuseops.InodeAttributes{
+			Uid:  uid,
+			Gid:  gid,
+			Mode: dirMode,
+		},
+		implicitDirs,
+		enableNonexistentTypecache,
+		typeCacheTTL,
+		&t.bucket,
+		&t.clock,
+		&t.clock)
+	t.dh = fs.NewDirHandle(in, implicitDirs)
+}
+
+func (t *DirHandleTest) resetDirHandle() {
+	t.dh = nil
+}
+
+// Directory Structure Used
+// foo   --Directory
+// fetchEntriesAsync will return 0 entries for empty directory
+func (t *DirHandleTest) FetchAsyncEntries_EmptyDir() {
+	t.createDirHandle(false, false, dirInodeName)
+	t.dh.FetchEntriesAsync(fuseops.RootInodeID, true)
+
+	AssertEq(0, len(t.dh.Entries))
+	AssertEq(true, t.dh.EntriesValid)
+	t.resetDirHandle()
+}
+
+// Directory Structure Used
+// foo       --Directory
+// foo/bar   --File
+// fetchEntriesAsync will return 1 entry for directory with 1 file
+func (t *DirHandleTest) FetchAsyncEntries_NonEmptyDir() {
+	contents := "Non-empty dir"
+	filePath := path.Join(dirInodeName, fileUnderDir)
+	_, err := t.bucket.CreateObject(
+		t.ctx,
+		&gcs.CreateObjectRequest{
+			Name:     filePath,
+			Contents: strings.NewReader(contents),
+		})
+	AssertEq(nil, err)
+
+	t.createDirHandle(false, false, dirInodeName)
+	t.dh.FetchEntriesAsync(fuseops.RootInodeID, true)
+
+	AssertEq(1, len(t.dh.Entries))
+	AssertEq(fileUnderDir, t.dh.Entries[0].Name)
+	AssertEq(true, t.dh.EntriesValid)
+	t.resetDirHandle()
+}
+
+// Directory Structure Used
+// foo              --Directory
+// foo/baz          --Implicit Directory
+// foo/baz/bar      --file
+// fetchEntriesAsync will return 1 entry for implicit directory if flag is set to true else 0
+func (t *DirHandleTest) FetchAsyncEntries_ImplicitDir() {
+	contents := "Implicit dir"
+	filePath := path.Join(dirInodeName, path.Join(implicitDirName, fileUnderDir))
+	_, err := t.bucket.CreateObject(
+		t.ctx,
+		&gcs.CreateObjectRequest{
+			Name:     filePath,
+			Contents: strings.NewReader(contents),
+		})
+	AssertEq(nil, err)
+
+	//implicit-dirs flag set to true
+	t.createDirHandle(true, false, dirInodeName)
+	t.dh.FetchEntriesAsync(fuseops.RootInodeID, true)
```

Comment:
dont merge two test cases in one . split it. you can create private functions to reuse the logic